src/main/java/com/commercetools/sync/categories/CategoryTypeDiff.java

```diff
@@ -0,0 +1,218 @@
+
+        if (Objects.equals(existingCustomFieldsTypeKey, newCustomFieldsDraftTypeKey)) {
+            if (existingCustomFieldsTypeKey == null && newCustomFieldsDraftTypeKey == null) {
+                // TODO: LOG THIS AS AN ERROR: "CUSTOM TYPE FOR BOTH EXISTING AND NEW CATEGORY IS NOT SET"
+                // TODO: wrap all processing results in a ProcessResult wrapper class.
+                return Collections.emptyList();
+            }
+            if (newCustomFieldsDraftJsonMap == null) {
+                // TODO: LOG THIS AS AN ERROR: "CUSTOM TYPE WITH NO CUSTOM FIELDS"
+                return Collections.emptyList();
+            }
+            // New and existing category's custom fields are set. So we should calculate update actions for the
+            // the fields of both.
+            return buildSetCustomFieldsActions(existingCustomFieldsJsonMap, newCustomFieldsDraftJsonMap);
+        } else {
+            return Collections.singletonList(
+                    SetCustomType.ofTypeKeyAndJson(
+                            newCustomFieldsDraftTypeKey, newCustomFieldsDraftJsonMap)
+            );
+        }
+    }
+
+    /**
+     * Compares two {@link Map<String, JsonNode>} representing a map of the custom field name to the JSON representation
+     * of the value of the corresponding custom field. It returns a {@link List<UpdateAction<Category>>} as a result.
+     * If no update action is needed an empty {@link List<UpdateAction<Category>>} is returned.
+     * <p>
+     * <p>
+     * An update action will be added to the result list in the following cases:-
+     * <p>
+     * 1. A custom field value is changed.
+     * 2. A custom field value is removed.
+     * 3. A new custom field value is added.
+     * <p>
+     * <p>
+     * <p>
+     * An update action will <bold>not</bold> be added to the result list in the following cases:-
+     * 1. Custom fields are both empty.
+     * 2. Custom field JSON values have different ordering.
+     * 3. Custom field values are identical.
+     *
+     * @param existingCustomFields the existing category's custom fields map of JSON values.
+     * @param newCustomFields      the new category's custom fields map of JSON values.
+     * @return a list that contains all the update actions needed, otherwise an empty list if no update actions are needed.
+     */
+    @Nonnull
+    static List<UpdateAction<Category>> buildSetCustomFieldsActions(
+            @Nonnull final Map<String, JsonNode> existingCustomFields,
+            @Nonnull final Map<String, JsonNode> newCustomFields) {
+        final List<UpdateAction<Category>> newOrModifiedCustomFieldsActions =
+                buildNewOrModifiedCustomFieldsActions(existingCustomFields, newCustomFields);
+        final List<UpdateAction<Category>> removedCustomFieldsActions =
+                buildRemovedCustomFieldsActions(existingCustomFields, newCustomFields);
+        return Stream.concat(newOrModifiedCustomFieldsActions.stream(),
```

Comment:
@lojzatran but actually you can't iterate once, you have to iterate two times once on the old list and once on the new list. To be able to know what's removed and what's added.